checker/src/org/checkerframework/checker/index/upperbound/UpperBoundStore.java

```diff
@@ -0,0 +1,337 @@
+    }
+
+    public UpperBoundStore(UpperBoundAnalysis analysis, boolean sequentialSemantics) {
+        super(analysis, sequentialSemantics);
+    }
+
+    public UpperBoundStore(
+            UpperBoundAnalysis analysis, CFAbstractStore<CFValue, UpperBoundStore> other) {
+        super(other);
+    }
+
+    @Override
+    public void updateForMethodCall(
+            MethodInvocationNode n, AnnotatedTypeFactory factory, CFValue val) {
+        super.updateForMethodCall(n, factory, val);
+
+        ExecutableElement elt = n.getTarget().getMethod();
+
+        if (!isSideEffectFree(factory, elt)) {
+            List<? extends Element> enclosedElts =
+                    ElementUtils.enclosingClass(elt).getEnclosedElements();
+            List<AnnotatedTypeMirror> enclosedTypes = new ArrayList<AnnotatedTypeMirror>();
+
+            findEnclosedTypes(enclosedTypes, enclosedElts);
+
+            // Should include all method calls and non-array references, but not the name of the method.
+            clearFromStore("", n, CheckController.NONFINAL_REFS_AND_METHOD_CALLS_METHOD);
+            checkForRemainingAnnotations(
+                    enclosedTypes, "", n, CheckController.NONFINAL_REFS_AND_METHOD_CALLS_METHOD);
+        }
+    }
+
+    @Override
+    public void updateForAssignment(Node n, CFValue val) {
+        // Do reassignment things here.
+        super.updateForAssignment(n, val);
+
+        // This code determines the list of dependences in types that are to be invalidated
+        CheckController checkController = null;
+
+        if (n.getType().getKind() == TypeKind.ARRAY) {
+            if (n instanceof LocalVariableNode) {
+                // Do not warn about assigning to a final variable. javac handles this.
+                if (!ElementUtils.isFinal(((LocalVariableNode) n).getElement())) {
+                    checkController = CheckController.NAME_ONLY;
+                }
+            }
+            if (n instanceof FieldAccessNode) {
+                // Do not warn about assigning to a final field. javac handles this.
+                if (!ElementUtils.isFinal(((FieldAccessNode) n).getElement())) {
+                    checkController = CheckController.NAME_AND_METHOD_CALLS;
+                }
+            }
+        } else {
+            if (n instanceof FieldAccessNode) {
+                if (!n.getType().getKind().isPrimitive()) {
+                    if (!ElementUtils.isFinal(((FieldAccessNode) n).getElement())) {
+                        checkController = CheckController.NONFINAL_REFS_AND_METHOD_CALLS;
+                    }
+                }
+            }
+        }
+
+        // Find all possibly-invalidated types
+        if (checkController != null) {
+
+            Element elt;
+
+            // So that assignments into arrays are treated correctly, as well as type casts
+            Node oldN = n;
+            while (n instanceof TypeCastNode || n instanceof ArrayAccessNode) {
```

Comment:
If `n` is not a FieldAccessNode nor a LocalVariableNode, then `checkController` is null. So this condition is always false.